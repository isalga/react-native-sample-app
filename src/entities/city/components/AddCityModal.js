// @flow
import type { City } from '../../types'
import React from 'react'
import { View, Text, Button, TouchableOpacity, Modal, StyleSheet } from 'react-native'
import Autocomplete from 'react-native-autocomplete-input'
import { connect } from 'react-redux'
import { addCity } from '../actions'
import { getAddedCities, getNotAddedCities } from '../selectors'

type Props = {
  addCity: (id: string) => void,
  toggleVisibility: () => void,
  cities: Object,
  visible: boolean
}

type State = {
  query: string
}

export class AddCityModal extends React.Component<Props, State> {

  static defaultProps = {
    addCity: () => {},
    toggleVisibility: () => {},
    cities: [],
    visible: false
  }

  state = {
    query: ''
  }

  onAddCityModalItemPress = (id: string) => () => {
    this.props.addCity(id)
    this.props.toggleVisibility()
  }

  findCities(query: string) {
    if (query === '') {
      return []
    }
    const regex = new RegExp(`${query.trim()}`, 'i')
    return this.props.cities.filter(city => city.displayName.search(regex) >= 0)
  }

  render() {
    const { query } = this.state
    const data = this.findCities(query)

    return (
        <Modal
          animationType='slide'
          transparent={false}
          visible={this.props.visible}
          onRequestClose={this.props.toggleVisibility}>
            <View style={styles.container}>
              <Button
                onPress={this.props.toggleVisibility}
                title='Close'
                accessibilityLabel='Close modal'/>
              <Autocomplete
                data={data}
                defaultValue={query}
                onChangeText={text => this.setState({ query: text })}
                renderItem={city => (
                  <TouchableOpacity onPress={this.onAddCityModalItemPress(city.id)}>
                    <Text>{city.displayName}</Text>
                  </TouchableOpacity>
                )}
              />
            </View>
        </Modal>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    marginTop: 30
  }
})

export default connect(
  (state) => ({
    cities: getNotAddedCities(state),
  }),
  (dispatch) => ({
    addCity: (id) => {
      dispatch(addCity(id))
    }
  })
)(AddCityModal)
