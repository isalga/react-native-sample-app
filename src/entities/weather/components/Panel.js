// @flow
import type { Weather } from '../../types'
import React from 'react'
import { View, Text, StyleSheet } from 'react-native'
import { colors } from '../../../styles'

type Props = {
  weather: ?Weather,
  title: ?string
}

const WeatherPanel = ({ weather, title }: Props) => {
  if (!weather)
    return (
      <View style={styles.container}>
        <Text>Uuups... information was not found</Text>
      </View>
    )
  else
    return (
      <View style={styles.container}>
        <Text>{ title }</Text>
        <Text>temperature: { weather.temperature }</Text>
        <Text>wind: { weather.windDirection } / { weather.windSpeed }</Text>
        <Text>humidity: { weather.humidity }</Text>
        <Text>precipitation: { weather.precipitation }</Text>
      </View>
    )
}

const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    padding: 20,
    backgroundColor: colors.widgetBackground,
    borderRadius: 4,
    borderWidth: 1,
    borderColor: colors.border
  }
})

export default WeatherPanel
