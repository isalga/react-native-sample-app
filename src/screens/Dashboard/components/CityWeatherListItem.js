// @flow
import type { City, Weather } from '../../../entities/types'
import React from 'react'
import { View, Text, TouchableOpacity, StyleSheet } from 'react-native'
import { Icon } from 'react-native-elements'
import { connect } from 'react-redux'
import { loadStart as loadWeatherStart } from '../../../entities/weather/actions'
import { setSelectedId, removeCity } from '../../../entities/city/actions'

type Props = {
  city: ?City,
  weather: ?Weather,
  selectCity: (id: string) => void,
  removeCity: (id: string) => void,
  loadWeather: (city: City) => void,
  navigate: (screen: string) => void
}

export class CityWeatherListItem extends React.Component<Props> {

  static defaultProps = {
    loadWeather: () => {},
    selectCity: () => {},
    removeCity: () => {},
    navigate: () => {}
  }

  componentDidMount() {
    if (!this.props.weather && this.props.city)
      this.props.loadWeather(this.props.city)
  }

  onPressItem = () => {
    if (this.props.city) {
      this.props.selectCity(this.props.city.id)
      this.props.navigate('CityWeather')
    }
  }

  onRemoveItem = () => {
    if (this.props.city) {
      this.props.removeCity(this.props.city.id)
    }
  }

  render() {
    const { city, weather } = this.props
    if (!city || typeof city != 'object')
      return null

    return (
      <View style={styles.listItem}>
        <TouchableOpacity
          onPress={this.onPressItem}>
          <Text>
            { city.displayName } - { weather ? weather.temperature : '...loading' }
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          onPress={this.onRemoveItem}>
          <Icon name='delete'/>
        </TouchableOpacity>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  listItem: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    padding: 10,
    height: 44
  },
})

export default connect(
  (state, { city }) => ({
    weather: state.weather[city.id]
  }),
  (dispatch) => ({
    selectCity: (id) => {
      dispatch(setSelectedId(id))
    },
    removeCity: (id) => {
      dispatch(removeCity(id))
    },
    loadWeather: (city) => {
      dispatch(loadWeatherStart(city))
    }
  })
)(CityWeatherListItem)
