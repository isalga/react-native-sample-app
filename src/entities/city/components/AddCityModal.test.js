import React from 'react'
import ReactDOM from 'react-dom'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { AddCityModal } from '../components/AddCityModal'

Enzyme.configure({ adapter: new Adapter() })
jest.mock('react-native-autocomplete-input', () => () => null)

const Abu_Dhabi = {
  id: '1',
  displayName: 'Abu Dhabi',
  name: 'Abu_Dhabi',
  country: 'united_arab_Emirates'
}
const Barcelona = {
  id: '2',
  displayName: 'Barcelona',
  name: 'Barcelona',
  country: 'spain'
}

const cities = [Abu_Dhabi, Barcelona]

describe('AddCityModal component', () => {

  it('should call props.addCity onAddCityModalItemPress', () => {
    const addCity = jest.fn()
    const wrapper = Enzyme.shallow(
      <AddCityModal
        cities={cities}
        toggleVisibility={() => {}}
        addCity={addCity}
        visible={true}/>)

    wrapper.instance().onAddCityModalItemPress('1')()

    expect(addCity.mock.calls.length).toEqual(1)
  })

  it('should render correctly', () => {
    const wrapper = Enzyme.shallow(
      <AddCityModal
        cities={cities}
        toggleVisibility={() => {}}
        addCity={() => {}}
        visible={true}/>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should render hidded element', () => {
    const wrapper = Enzyme.shallow(
      <AddCityModal
        cities={cities}
        toggleVisibility={() => {}}
        addCity={() => {}}
        visible={false}/>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should not break', () => {
    const wrapper = Enzyme.shallow(
      <AddCityModal/>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
