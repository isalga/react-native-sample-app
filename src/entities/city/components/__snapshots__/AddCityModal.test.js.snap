// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCityModal component should not break 1`] = `
<Component
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={false}
  visible={false}
>
  <View
    style={
      Object {
        "marginTop": 30,
      }
    }
  >
    <Button
      accessibilityLabel="Close modal"
      onPress={[Function]}
      title="Close"
    />
    <Component
      data={Array []}
      defaultValue=""
      onChangeText={[Function]}
      renderItem={[Function]}
    />
  </View>
</Component>
`;

exports[`AddCityModal component should render correctly 1`] = `
<Component
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={false}
  visible={true}
>
  <View
    style={
      Object {
        "marginTop": 30,
      }
    }
  >
    <Button
      accessibilityLabel="Close modal"
      onPress={[Function]}
      title="Close"
    />
    <Component
      data={Array []}
      defaultValue=""
      onChangeText={[Function]}
      renderItem={[Function]}
    />
  </View>
</Component>
`;

exports[`AddCityModal component should render hidded element 1`] = `
<Component
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={false}
  visible={false}
>
  <View
    style={
      Object {
        "marginTop": 30,
      }
    }
  >
    <Button
      accessibilityLabel="Close modal"
      onPress={[Function]}
      title="Close"
    />
    <Component
      data={Array []}
      defaultValue=""
      onChangeText={[Function]}
      renderItem={[Function]}
    />
  </View>
</Component>
`;
